src/shamir.rs

```diff
@@ -0,0 +1,419 @@
+                .map(|(lhs, rhs)| lhs + rhs)
+                .collect::<Vec<_>>();
+
+            let sum_secret = SecretSharing::reconstruct(&sum_shares, lc).unwrap();
+            assert_eq!(expected, sum_secret);
+        }
+
+        let k = NonZeroU8::new(2).unwrap();
+        let n = NonZeroU8::new(3).unwrap();
+        let shamir = SecretSharing::new(k, n).unwrap();
+        let lc = LagrangePolynomial::<Fp31>::new(n).unwrap();
+
+        check_addition(&shamir, &lc, 42, 24);
+        check_addition(&shamir, &lc, 0, 0);
+        check_addition(&shamir, &lc, Fp31::PRIME, 2);
+        check_addition(&shamir, &lc, 0, Fp31::PRIME);
+    }
+
+    #[test]
+    fn fails_if_not_enough_shares() {
+        let n = NonZeroU8::new(3).unwrap();
+        let shamir = SecretSharing::new(NonZeroU8::new(2).unwrap(), n).unwrap();
+
+        let shares = shamir.split(Fp31::from(42_u128), thread_rng());
+        assert!(matches!(
+            SecretSharing::reconstruct(&shares[0..1], &LagrangePolynomial::new(n).unwrap()),
+            Err(Error::BadPolynomial {
+                polynomial_degree: 3,
+                points_count: 1
+            })
+        ));
+    }
+
+    #[test]
+    fn fails_if_not_enough_coefficients() {
+        let k = NonZeroU8::new(2).unwrap();
+        let n = NonZeroU8::new(3).unwrap();
+        let shamir = SecretSharing::new(k, n).unwrap();
+
+        let shares = shamir.split(Fp31::from(42_u128), thread_rng());
+        assert!(matches!(
+            SecretSharing::reconstruct(&shares, &LagrangePolynomial::new(k).unwrap()),
+            Err(Error::BadPolynomial {
+                polynomial_degree: 2,
+                points_count: 3
+            })
+        ));
+    }
+
+    #[test]
+    fn can_reject_bad_k_and_n() {
+        for k in [1_u8, 2, 3] {
+            let n = NonZeroU8::new(max(1, k - 1)).unwrap();
+            let k = NonZeroU8::new(k).unwrap();
+            let r = SecretSharing::new(k, n);
+            assert!(matches!(r, Err(Error::BadSharingScheme { .. })));
+        }
+    }
+
+    #[test]
+    fn fails_to_reconstruct_secret_with_duplicated_shares() {
+        let (k, n) = (NonZeroU8::new(2).unwrap(), NonZeroU8::new(3).unwrap());
+        let rng = StepRng::new(1, 1);
+        let secret = Fp31::from(5_u128);
+
+        let sharing = SecretSharing::new(k, n).unwrap();
+
+        let mut shares = sharing.split(secret, rng);
+        shares[2] = shares[1].clone();
+
+        assert_ne!(
+            secret,
+            SecretSharing::reconstruct(&shares, &LagrangePolynomial::new(n).unwrap()).unwrap()
+        );
+    }
+
+    #[test]
+    fn fails_to_reconstruct_secret_with_forged_shares() {
+        let (k, n) = (NonZeroU8::new(2).unwrap(), NonZeroU8::new(3).unwrap());
+        let rng = StepRng::new(1, 1);
+        let secret = Fp31::from(5_u128);
+
+        let sharing = SecretSharing::new(k, n).unwrap();
+
+        let mut shares = sharing.split(secret, rng);
+        shares[2] = Share { y: shares[1].y };
```

Comment:
What exactly is going on here? What do you mean by "forged"?

Also, if k=2, you should only need two shares to reconstruct I think.

Reply:
yes, this test ensures that reconstruction fails if we give it bad input such as S0, S1, S1